Fix compilation with -Wpedantic on GCC

Mixaill opened the pull request:
Description
-----------

This PR does the following
- fixes `empty declaration of 'enum' type does not redeclare tag` error in the `FreeRTOS_IPv4.h` file.
- adds `-Wpedantic` to the GCC compilation flags
   - adds `-Wno-pedantic` for `FreeRTOS_Sockets.c` file in case you are using GCC to suppress warning

The main reason is that `FreeRTOS_IPv4.h` is a public header and in case an application uses `-Wpedantic -Werror` to compile, it will not be able to use FreeRTOS-Plus-TCP.

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

moninom1 commented on source/FreeRTOS_Sockets.c:
```diff
@@ -2900,7 +2900,14 @@ BaseType_t FreeRTOS_setsockopt( Socket_t xSocket,
                         /* More details at: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/MISRA.md#rule-111 */
                         /* coverity[misra_c_2012_rule_11_8_violation] */
```
Can we please move these 2 lines inside ifdef, the MISRA suppression should be immediately followed by the MISRA issue.

moninom1 replied:
Can we please move these 2 lines[2901,2902] inside ifdef, the MISRA suppression should be immediately followed by the MISRA issue.

moninom1 commented on source/include/FreeRTOS_IPv4.h:
```diff
@@ -43,9 +43,17 @@
-enum eFrameProcessingResult;
 struct xIP_PACKET;
 
+#ifdef __GNUC__
```
Why is this needed, i see that by adding -Wpedantic to the GCC compilation flags only FreeRTOS_Sockets.c issue is seen.

Mixaill replied:
It gives an error on GCC 13.2

```
/opt/arm/gcc-13.2/bin/arm-none-eabi-gcc 

-mcpu=cortex-m7 -mfpu=fpv5-d16 -mfloat-abi=hard -mthumb -mlong-calls -mlittle-endian -Wall -Wextra -Werror -Wpedantic -Wno-psabi -fstack-usage -ffunction-sections -fdata-sections -fno-omit-frame-pointer -fvar-tracking -fvar-tracking-assignments -O0 -DDEBUG  -ggdb3 -gdwarf-5 -std=gnu2x -fdiagnostics-color=always -Wno-array-bounds -Wno-strict-aliasing -Wno-unused-but-set-variable

...
In file included from .../FreeRTOS_IP.h:496,
                 from .../FreeRTOS_DNS_Networking.h:30,
                 from .../FreeRTOS_DNS_Networking.c:36:
FreeRTOS_IPv4.h:50:10: error: empty declaration of 'enum' type does not redeclare tag [-Werror=pedantic]
   50 |     enum eFrameProcessingResult;
``

moninom1 commented:
Hi @Mixaill 

Can you please help in updating on the above review comments.

Mixaill commented:
@moninom1

aggarg commented:
Thank you for your contribution. I have the following suggestions:

* Warning from `source/FreeRTOS_Sockets.c` - We do not want to add compiler specific stuff in the common source files as those are built with multiple compilers. The recommended way to suppress the warning is to suppress it in the build system (preferably only for this file) if you enable pedantic option.

* Warning from `source/include/FreeRTOS_IPv4.h` - The correct way to address this warning is to remove the forward declaration of `eFrameProcessingResult` and move the following 2 function declarations to `source/include/FreeRTOS_IPv4_Private.h`:
  ```
  /* The function 'prvAllowIPPacket()' checks if a packets should be processed. */
  enum eFrameProcessingResult prvAllowIPPacketIPv4( const struct xIP_PACKET * const pxIPPacket,
                                                    const struct xNETWORK_BUFFER * const pxNetworkBuffer,
                                                    UBaseType_t uxHeaderLength );
  
  /* Check if the IP-header is carrying options. */
  enum eFrameProcessingResult prvCheckIP4HeaderOptions( struct xNETWORK_BUFFER * const pxNetworkBuffer );
  ```
  @tony-josi-aws has verified that this changes fix the warning - https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/commit/129b4ceb4220f5141615afeafa3c79ab998b1005.

Would you like to make these changes or would you like us to?

Mixaill commented:
@aggarg 

* replaced changes in `FreeRTOS_Sockets.c` with `-Wno-pedantic` for this file
* replaced changes in `FreeRTOS_IPv4.h` with commits from @tony-josi-aws
